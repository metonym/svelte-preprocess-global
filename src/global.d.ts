import type { Element } from "svelte/types/compiler/interfaces";
import type { PreprocessorGroup } from "svelte/types/compiler/preprocess";

declare global {
  interface SveltePreprocessor<PreprocessorType, Options = any> {
    (options?: Options): Pick<PreprocessorGroup, PreprocessorType>;
  }

  interface BaseNode {
    type: string;
    name: string;
    start: number;
    end: number;
  }

  interface DeclarationNode extends BaseNode {
    type: "Declaration";
    property: string;
    value: {
      children: IdentifierNode[];
    };
  }

  interface AtruleNode extends BaseNode {
    type: "Atrule";
    prelude: BaseNode & {
      children: IdentifierNode[];
    };
  }

  interface ClassSelectorNode extends BaseNode {
    type: "ClassSelector";
    children: any[];
  }

  interface IdSelectorNode extends BaseNode {
    type: "IdSelector";
  }

  interface SelectorNode extends BaseNode {
    type: "Selector";
    children: ClassSelectorNode[];
  }

  interface IdentifierNode extends BaseNode {
    type: "Identifier";
    name: string;
  }

  interface AttributeSelectorNode extends Omit<BaseNode, "name"> {
    type: "AttributeSelector";
    name: IdentifierNode;
  }

  type AstNode =
    | Element
    | AttributeSelectorNode
    | ClassSelectorNode
    | SelectorNode
    | IdSelectorNode
    | DeclarationNode
    | AtruleNode;
}
