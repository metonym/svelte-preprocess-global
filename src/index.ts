import { parse, walk } from "svelte/compiler";
import MagicString from "magic-string";
import type { PreprocessorGroup } from "svelte/types/compiler/preprocess";
import type { Element } from "svelte/types/compiler/interfaces";

interface BaseNode {
  type: string;
  name: string;
  start: number;
  end: number;
}

interface DeclarationNode extends BaseNode {
  type: "Declaration";
  property: string;
  value: {
    children: IdentifierNode[];
  };
}

interface AtruleNode extends BaseNode {
  type: "Atrule";
  prelude: BaseNode & {
    children: IdentifierNode[];
  };
}

interface ClassSelectorNode extends BaseNode {
  type: "ClassSelector";
  children: any[];
}

interface IdSelectorNode extends BaseNode {
  type: "IdSelector";
}

interface SelectorNode extends BaseNode {
  type: "Selector";
  children: ClassSelectorNode[];
}

interface IdentifierNode extends BaseNode {
  type: "Identifier";
  name: string;
}

interface AttributeSelectorNode extends Omit<BaseNode, "name"> {
  type: "AttributeSelector";
  name: IdentifierNode;
}

type Node =
  | Element
  | AttributeSelectorNode
  | ClassSelectorNode
  | SelectorNode
  | IdSelectorNode
  | DeclarationNode
  | AtruleNode;

type StyleKey = ClassSelectorNode["type"] | IdSelectorNode["type"] | AttributeSelectorNode["type"];

const PREFIX: Record<StyleKey, (selector: string) => string> = {
  ClassSelector: (selector) => "." + selector,
  IdSelector: (selector) => "#" + selector,
  AttributeSelector: (selector) => "[" + selector + "]",
};

const useGlobal = (selector: string) => ":global(" + selector + ")";

const GLOBAL_KEYFRAME = "-global-";

export function global(): Pick<PreprocessorGroup, "markup"> {
  return {
    markup({ content, filename }) {
      if (filename && /(node_modules|.svelte-kit)/.test(filename)) return;

      const s = new MagicString(content);
      const ids = new Set();
      const class_names = new Set();
      const data_attributes = new Set();
      const keyframes = new Set();

      const apply_global = ({
        node,
        parent,
        selector,
      }: {
        node: IdSelectorNode | ClassSelectorNode | AttributeSelectorNode;
        parent: Node;
        selector: string;
      }) => {
        let start = node.start;
        let end = node.end;
        let value = useGlobal(PREFIX[node.type](selector));

        if ((parent as SelectorNode).children.length > 1) {
          start = parent.start;
          end = parent.end;
          value = useGlobal(content.slice(start, end));
        }

        s.overwrite(start, end, value);
      };

      walk(parse(content), {
        enter(node: Node, parent: Node) {
          if (node.type === "InlineComponent") {
            node.attributes.forEach(({ name, value }) => {
              const raw: string = value[0]?.raw;

              if (name === "id") {
                ids.add(raw);
              } else if (name === "class") {
                raw.split(" ").forEach(class_names.add.bind(class_names));
              } else if (/^data-/.test(name)) {
                data_attributes.add(name);
              }
            });
          }

          if (
            node.type === "Declaration" &&
            (node.property === "animation" || node.property === "animation-name")
          ) {
            keyframes.add(node.value.children[0]?.name);
          }

          if (node.type === "ClassSelector" && class_names.has(node.name)) {
            apply_global({ node, parent, selector: node.name });
          }

          if (node.type === "IdSelector" && ids.has(node.name)) {
            apply_global({ node, parent, selector: node.name });
          }

          if (node.type === "AttributeSelector" && data_attributes.has(node.name.name)) {
            apply_global({ node, parent, selector: node.name.name });
          }

          if (node.type === "Atrule" && /keyframes/.test(node.name)) {
            const keyframe = node.prelude.children[0].name;
            if (keyframes.has(keyframe)) {
              s.overwrite(node.prelude.start, node.prelude.end, GLOBAL_KEYFRAME + keyframe);
            }
          }
        },
      });

      return {
        code: s.toString(),
        map: s.generateMap({ file: filename, includeContent: true }),
      };
    },
  };
}
