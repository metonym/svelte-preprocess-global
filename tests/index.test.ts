import { expect, test } from "vitest";
import * as API from "../src";
import type { Processed } from "svelte/types/compiler/preprocess";

test("API", () => {
  expect(Object.keys(API)).toMatchInlineSnapshot(`
    [
      "global",
    ]
  `);
});

const preprocess = (content: string) => (API.global().markup?.({ content }) as Processed).code;

test("no styles", () => {
  expect(preprocess(`<Component id="my-id" class="class1 class2" />`)).toMatchSnapshot();
});

test("single id, class", () => {
  expect(
    preprocess(`<Component id="my-id" class="class1 class2" />

<h1>Hello world</h1>

<style>
  #my-id { outline: 1px solid blue; }

  .class1 { color: red; }

  h1 { font-size: 2rem; }
</style>`)
  ).toMatchSnapshot();
});

test("styles first, multiple selectors", () => {
  expect(
    preprocess(`<style>
  .class1, .class2 { color: red; }

  .class2 { outline: 1px solid red; }

  .class2 div, #h1, .not-a-class { color: blue; }

  #my-id > div, #h1, .not-a-class { color: blue; }
</style>

<Component id="my-id" class="class1 class2" />`)
  ).toMatchSnapshot();
});

test("single id, class", () => {
  expect(
    preprocess(`<Component id="my-id" class="class1 class2" />

<style>
  #my-id { outline: 1px solid blue; }

  .class1 { color: red; }
</style>`)
  ).toMatchSnapshot();
});

test("data attributes", () => {
  expect(
    preprocess(`<style>
  [data-attr] { color: red; }

  [data-attr2] div, data-value { color: blue; }
</style>

<Component data-attr data-attr2="value" />`)
  ).toMatchSnapshot();
});

test("readme example", () => {
  expect(
    preprocess(`<Component id="component" data-component class="bg-blue" />

<style>
  #component {
    color: red;
  }

  [data-component] {
    outline: 1px solid white;
  }

  .bg-blue {
    background: blue;
  }
</style>`)
  ).toMatchSnapshot();
});

test("keyframes", () => {
  expect(
    preprocess(`<Component class="animate animate-name transition transition-name" />

<style>
  .animate {
    animation: fade 1.5s linear infinite;
  }

  .animate-name {
    animation-name: fade-out;
  }

  @keyframes fade {
    50% {
      opacity: 0;
    }
  }

  @-webkit-keyframes fade {
    50% {
      opacity: 0;
    }
  }

  @keyframes fade-out {
    50% {
      opacity: 0;
    }
  }
</style>`)
  ).toMatchSnapshot();
});
