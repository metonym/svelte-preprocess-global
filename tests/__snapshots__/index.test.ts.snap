// Vitest Snapshot v1

exports[`data attributes 1`] = `
"<style>
  :global([data-attr]) { color: red; }

  :global([data-attr2] div), data-value { color: blue; }
</style>

<Component data-attr data-attr2=\\"value\\" />"
`;

exports[`keyframes 1`] = `
"<Component class=\\"animate animate-name transition transition-name\\" />

<style>
  :global(.animate) {
    animation: fade 1.5s linear infinite;
  }

  :global(.animate-name) {
    animation-name: fade-out;
  }

  @keyframes -global-fade {
    50% {
      opacity: 0;
    }
  }

  @-webkit-keyframes -global-fade {
    50% {
      opacity: 0;
    }
  }

  @keyframes -global-fade-out {
    50% {
      opacity: 0;
    }
  }
</style>"
`;

exports[`no styles 1`] = `"<Component id=\\"my-id\\" class=\\"class1 class2\\" />"`;

exports[`readme example 1`] = `
"<Component id=\\"component\\" data-component class=\\"bg-blue\\" />

<style>
  :global(#component) {
    color: red;
  }

  :global([data-component]) {
    outline: 1px solid white;
  }

  :global(.bg-blue) {
    background: blue;
  }
</style>"
`;

exports[`single id, class 1`] = `
"<Component id=\\"my-id\\" class=\\"class1 class2\\" />

<style>
  :global(#my-id) { outline: 1px solid blue; }

  :global(.class1) { color: red; }
</style>"
`;

exports[`single id, class 2`] = `
"<Component id=\\"my-id\\" class=\\"class1 class2\\" />

<style>
  :global(#my-id) { outline: 1px solid blue; }

  :global(.class1) { color: red; }
</style>"
`;

exports[`styles first, multiple selectors 1`] = `
"<style>
  :global(.class1), :global(.class2) { color: red; }

  :global(.class2) { outline: 1px solid red; }

  :global(.class2 div), #h1, .not-a-class { color: blue; }

  :global(#my-id > div), #h1, .not-a-class { color: blue; }
</style>

<Component id=\\"my-id\\" class=\\"class1 class2\\" />"
`;
